### ISSUE TITLE
[BUG] Shapelet Transform not normalising constant series

### ISSUE THREAD

**Issue by TonyBagnall** [aeon-toolkit/aeon#952]
### Describe the bug

this is a logical bug not a runtime error. Shapelets are normalised. Constant shapelets just subtract the mean. However, if they are constant (zero variance),  you can have non zero mean shapelets, for example
![image](https://github.com/aeon-toolkit/aeon/assets/9594042/1a49979c-38fc-4bb0-9b97-393da44c9c1d)

this will probably not make a difference to the quality measure, a constant offset will not alter the info gain, but its not imo actually correct, should zero it. From the code though, I cant see why this is happening, see below. 

### Steps/Code to reproduce the bug

to recreate

```python 
from aeon.datasets import load_classification
trainX, trainY, metaTrain = load_classification("ScreenType",split="TRAIN")
from aeon.transformations.collection.shapelet_based import RandomShapeletTransform
st = RandomShapeletTransform(random_state=1)
st.fit(trainX, trainY)
```
then inspect the shapelets in st, you will see ones all 1 shapelets. 

Looking into it, I cant see how this happens, unless this is not always called: 
```python
@njit(fastmath=True, cache=True)
def z_normalise_series(X: np.ndarray) -> np.ndarray:
    s = stats.std(X)
    if s > 0:
        arr = (X - stats.mean(X)) / s
    else:
        arr = X - stats.mean(X)
    return arr
```
        arr = X - stats.mean(X)
it should zero a constant series

seems to always be called here
```python
        shapelet = z_normalise_series(
            X[inst_idx][channel][position : position + length]
        )
        sabs = np.abs(shapelet)
```
mysterious

### Expected results

constant shapelets should be all zeros

### Actual results

![image](https://github.com/aeon-toolkit/aeon/assets/9594042/eec09a09-e7a6-4d1e-914c-aae39d7b84aa)

**Comment by TonyBagnall** [comment]
easier way to reproduce 
```python
from aeon.datasets import load_classification
trainX, trainY, metaTrain = load_classification("ScreenType",split="TRAIN")
testX, testY, metaTest = load_classification("ScreenType",split="TRAIN")

from aeon.transformations.collection.shapelet_based import RandomShapeletTransform

st = RandomShapeletTransform(random_state=1, n_shapelet_samples=200)
trans1=st.fit_transform(trainX,trainY)
#st2 = RandomShapeletTransform(random_state=1, n_shapelet_samples=50000)
#trans2=st2.fit_transform(trainX,trainY)

print("Constant shapelet =",st.shapelets[8][6])
print("Constant shapelet =",st.shapelets[9][6])

print("Constant shapelet =",st.shapelets[8])
```

**Comment by baraline** [comment]
I add similar issues when using numba and fastmath options, sometime the std was not exactly 0 (e.g. 1e-7) due to fastmath rounding. One fix I did was to impose a STD_THRESHOLD (AEON_SIMSEARCH_STD_THRESHOLD = 1e-7 for similarity search), and test this equality to this instead of 0.

**Comment by TonyBagnall** [comment]
dont suppose we have a global tolerance defined anywhere?

**Comment by TonyBagnall** [comment]
where to put it ...

and looking at the above some of the code in distance_profiles/_commons.py seems a duplicate of that in utils/numba/general.py. Any chance we could rationalise it, I think utils/numba is a good central location?

**Comment by baraline** [comment]
> where to put it ...
> 
> and looking at the above some of the code in distance_profiles/_commons.py seems a duplicate of that in utils/numba/general.py. Any chance we could rationalise it, I think utils/numba is a good central location?

I'll take a look at it and remove the duplicates when I have sorted the recompilation thing for the refactoring. In terms of location for the threshold, why not in aeon/__init__.py ? Or define a file with all constant in utils, don't know what is best

**Comment by baraline** [comment]
To summarize the disucssion made on Slack, the problem might be simply linked to the way we compute the standard deviation in some cases:
In the case of the shapelet transform, when the online distance use this way of computing std, even without numba: 
```python
def _online_shapelet_std_py(subseq):
    length = subseq.shape[0]

    sum = 0.0
    sum2 = 0.0
    for i in subseq:
        sum += i
        sum2 += i * i

    mean = sum / length
    print((sum2 - mean * mean * length) / length)
    std = ((sum2 - mean * mean * length) / length)**0.5
    return std

assert not np.isnan(_online_shapelet_std_py(np.random.rand(1000)))
assert not np.isnan(_online_shapelet_std_py(np.ones(1000)*5.))
assert not np.isnan(_online_shapelet_std_py(np.ones(1000)*5.7))
assert not np.isnan(_online_shapelet_std_py(np.ones(1000)*5.6))
```
We sometime get NaN anyway, because we obtain a very small negative std instead of 0 or small positive. The NaN comes from trying to use sqrt on a negative value.

This case can thus be solved by checking that the variance is a very small positive value, such as 1e-11, before applying the square root. In use cases where we normalize, we can return a arrays of zeros bellow this threshold.
